### PR TITLE
all: shift all usages of ddtrace.Span to *tracer.Span

### PR DESCRIPTION
**Pull request by knusbaum**
### What does this PR do?

As part of our v2 migration, we want to stop using the `ddtrace.Span` interface and use a struct type instead, to allow us to expand the API as the tracer grows.

In order to do this, we need to:
- expose `tracer.span` as `tracer.Span`
- change everything to use `*tracer.Span` in place of `ddtrace.Span`
- Refactor `*tracer.Span` to *not* export member fields.

There are currently two commits in this PR, which can be viewed separately. The first one exposing `tracer.span` and the second one migrating its use.

### Reviewer's Checklist

- [ ] Changed code has unit tests for its functionality at or near 100% coverage.
- [ ] [System-Tests](https://github.com/DataDog/system-tests/) covering this feature have been added and enabled with the va.b.c-dev version tag.
- [ ] There is a benchmark for any new code, or changes to existing code.
- [ ] If this interacts with the agent in a new way, a system test has been added.
- [ ] Add an appropriate team label so this PR gets put in the right place for the release notes.

For Datadog employees:

- [ ] If this PR touches code that handles credentials of any kind, such as Datadog API keys, I've requested a review from `@DataDog/security-design-and-guidance`.
- [ ] This PR doesn't touch any of that.

Unsure? Have a question? Request a review!

### CODE REVIEW

**Comment by knusbaum** [comment]
This is ready for review. Tests are not going to pass here because the follow-on work of updating everything to use the new API is in #2409.

Please review this as if all tests pass, as they have been addressed in #2409.

The reason for separating these two PRs is to keep the diff small enough to be manageable.

In this PR, I eliminate the `ddtrace.Span` interface and do the work of moving types around to avoid dependency cycles, etc.

My plan is this:
* Get this PR reviewed
* Put up follow-on PRs to this branch (i.e. #2409) to finish the work, and allow reviewing the work piecemeal 
* Merge the follow-on PRs individually into this branch
* Do one final review of this branch before merging into v2-dev.

**Review by darccio — CHANGES_REQUESTED** [review]
Check the section "Unchanged files with check annotations" and `SetRequestMetadataTags(span ddtrace.Span, md map[string][]string)`.

Edit: in general, great work! It looks great. I look forward to see the next PRs.

**Comment by darccio** [comment]
@knusbaum I think my changes got us closer to the green CI but it's still failing in the system-tests: https://github.com/DataDog/dd-trace-go/actions/runs/7260335970/job/19779343227?pr=2408#step:5:570

**Review by darccio — APPROVED** [review]

File: ddtrace/tracer/tracer.go
Comment (nsrip-dd) [inline ddtrace/tracer/tracer.go:659]: I assume this is needed now because `SpanStart` unconditionally puts something in this field, right? But if I comment out this line, all the tests still pass. So I worry there is some untested behavior here.

Related, but is there a benefit to splitting the pprof label and execution trace task logic between these two span creation functions? GitHub won't let me comment on deleted lines, but I'm referring to the deleted block preceding this line. I think I prefer the way it was before where `startExecutionTracerTask` and `applyPPROFLabels` happen back to back. That way I find it's more clear that both things are trying to carry values through the same context. The context propagation logic is already tricky as it is (I struggle to wrap my head around it any time I need to work with it) and I worry it will become even more difficult to understand if it's spread out more in the code.

Comment (knusbaum) [reply ddtrace/tracer/tracer.go:659]: We're trying to split the `Span` logic out of the `Tracer` as much as possible, so we can make `Spans` independent of the tracer. This way we can use Spans with things like the mock tracer or other potential implementations.

Unfortunately, there is high coupling there. Some things are tracer-specific, including this pprof stuff. It would be nice to make this independent of the tracer entirely, but as you said, the logic is complicated and I didn't want to change it any more than absolutely necessary to achieve separation.

Please let me know if you see a better way to handle this.

> I assume this is needed now because SpanStart unconditionally puts something in this field, right? But if I comment out this line, all the tests still pass. So I worry there is some untested behavior here.

Correct. You're probably right too about missing coverage. I'll look into that.

Comment (nsrip-dd) [reply ddtrace/tracer/tracer.go:659]: > Correct. You're probably right too about missing coverage. I'll look into that.

I was able to exercise a difference with this test:

```go
func TestPprofLabels(t *testing.T) {
	if err := tracer.Start(
		tracer.WithProfilerCodeHotspots(false),
		tracer.WithProfilerEndpoints(false),
	); err != nil {
		t.Fatal(err)
	}
	defer tracer.Stop()
	pprof.Do(context.Background(), pprof.Labels("foo", "bar"), func(ctx context.Context) {
		wasteA(time.Second)
		var span *tracer.Span
		pprof.Do(ctx, pprof.Labels("foo", "baz"), func(ctx context.Context) {
			span, _ = tracer.StartSpanFromContext(ctx, "myoperation")
			wasteB(time.Second)
		})
		span.Finish()
		wasteC(time.Second)
	})
}
```

Here, `waste{A,B,C}` are just functions which waste the given amount of CPU time, so I can have some samples to distinguish which labels are active for the goroutine. I ran this with the CPU profiler enabled and looked at what labels are assigned to the `wasteC` samples. I would expect, with code hotspots and endpoint profiling disabled, that the `wasteC` samples have the label `"foo": "bar"`. Indeed, that's the case with this PR as-is and on main. If I comment out the `span.pprofCtxRestore = nil` case in this PR, the label for `wasteC` samples becomes `"foo": "baz"`. This helps convince me that this line is correct. Granted, this is a somewhat silly example. I don't know if a user would ever write code like this. But if we feel that it would be a good thing to have, I can turn it into a proper test (actually check the CPU profile output) as a followup.

Tangential, but I think as a user I might be surprised to find that with code hotspots and/or endpoint profiling _enabled_, the `wasteC` samples are tagged with `"foo": "baz"`. Naively reading the code, it wouldn't be obvious that the tracer functions are manipulating the labels at all. But, this PR is at least consistent with the current behavior, so... 🤷 The semantics of pprof labels don't really play well with these distributed tracing APIs, at least not nearly as well as execution tracer tasks do.

Our existing test suite passes otherwise, including traceproftest, so code hotspots & endpoint profiling seem to still work. The execution tracer tests here and in v2/contrib/database/sql pass as well (tried with Go 1.20). So it doesn't appear to be any more broken than it was before. With that in mind, I think I could tentatively approve this.

> Please let me know if you see a better way to handle this.

I don't have a better idea at the moment, unfortunately :/

If we want to feel more confident refactoring this code, I think we'd really benefit from nailing down the expected behavior with regard to context propagation, pprof labels, etc with good tests. I can tackle this, but I don't want to unduly block this work since I don't know when I'll have time to look into it. I do worry it will be even more difficult if we have to fix up two divergent versions of this code.

File: .github/workflows/smoke-tests.yml
Comment (RomainMuller) [inline .github/workflows/smoke-tests.yml:49]: Could the use of go workspaces have simplified this, maybe?

File: appsec/appsec.go
Comment (RomainMuller) [inline appsec/appsec.go:140]: That comment seems outdated (apparently, `tracer.Span` declares the `Root` method.

File: ddtrace/mocktracer/mockspan.go
Comment (RomainMuller) [inline ddtrace/mocktracer/mockspan.go:25]: Is this really necessary? This struct has a single pointer embed... of a type by the same leaf name.

Or is this supposed to be the new `MockSpan` type? In which case perhaps this needs renaming?

File: ddtrace/mocktracer/mockspan.go
Comment (RomainMuller) [inline ddtrace/mocktracer/mockspan.go:66]: I'd have allocated the map _after_ the nil-check, so that the map can be pre-allocated to the correct size.

File: ddtrace/mocktracer/mockspan.go
Comment (RomainMuller) [inline ddtrace/mocktracer/mockspan.go:62]: It's weird this method doesn't acquire the `s.RLock()`... is that an omission?

File: ddtrace/mocktracer/mockspan.go
Comment (RomainMuller) [inline ddtrace/mocktracer/mockspan.go:80]: It's weird that the other methods appear to expect/deal with a `nil` receiver, but this one doesn't... I'd argue either all need to gracefully deal with a `nil` receiver, or none should.

File: ddtrace/mocktracer/mockspan.go
Comment (RomainMuller) [inline ddtrace/mocktracer/mockspan.go:88]: Seems plausible that this could also be pre-allocated to the correct size upfront

File: ddtrace/mocktracer/mockspan.go
Comment (RomainMuller) [inline ddtrace/mocktracer/mockspan.go:99]: Looks like this _should_ deadlock, had accesses to `Tags` been adequately gated with `s.RLock()`...

Unless the `s.RLock()` means to guard only the `baggage`? If that is the case, why is the lock not on the `sc` (which hosts the baggage)?

File: ddtrace/mocktracer/mockspan.go
Comment (RomainMuller) [inline ddtrace/mocktracer/mockspan.go:126]: I'd have called this `Unwrap` to mimic other similar APIs that exist in other places, but that's also a decent name.

File: ddtrace/mocktracer/mockspancontext.go
Comment (RomainMuller) [inline ddtrace/mocktracer/mockspancontext.go:27]:
```suggestion
	span *tracer.Span
```

File: ddtrace/mocktracer/mocktracer.go
Comment (RomainMuller) [inline ddtrace/mocktracer/mocktracer.go:63]: This might be an unpopular opinion, but if this method returned `tracer.Tracer`, then the `var _ tracer.Tracer = (*mocktracer)(nil)` would become redundant.

File: ddtrace/mocktracer/mocktracer.go
Comment (RomainMuller) [inline ddtrace/mocktracer/mocktracer.go:77]: I find the semantics here a little uncanny... Why can we directly assign to `tracer.Testing` but we have to call a method to set (*waves hands*) `tracer.GlobalTracer`?

It also feels like `tracer.Testing` is only set after `tracer.SetGlobalTracer` is called, so why not have something like:
```suggestion
	tracer.SetGlobalTracer(&tracer.NoopTracer{}, false)
```

File: ddtrace/tracer/context.go
Comment (RomainMuller) [inline ddtrace/tracer/context.go:31]:
```suggestion
		return nil, false
```

File: ddtrace/tracer/context.go
Comment (RomainMuller) [inline ddtrace/tracer/context.go:38]:
```suggestion
	return nil, false
```

File: ddtrace/tracer/context.go
Comment (RomainMuller) [inline ddtrace/tracer/context.go:72]:
```suggestion
	return s, ContextWithSpan(ctx, s)
```

File: ddtrace/tracer/globaltracer.go
Comment (RomainMuller) [inline ddtrace/tracer/globaltracer.go:30]: This actually confirms what I was saying above... it feels like `Testing` could be passed as a boolean parameter to `SetGlobalTracer` and perhaps renamed to `stopPrevious` or something...

File: ddtrace/tracer/globaltracer.go
Comment (RomainMuller) [inline ddtrace/tracer/globaltracer.go:45]: I'd be somewhat partial to having this in a separate file (no biggie)

File: ddtrace/tracer/sampler.go
Comment (RomainMuller) [inline ddtrace/tracer/sampler.go:81]:
```suggestion
```

File: ddtrace/tracer/span.go
Comment (RomainMuller) [inline ddtrace/tracer/span.go:49]:
```suggestion
```

File: ddtrace/tracer/span.go
Comment (RomainMuller) [inline ddtrace/tracer/span.go:225]: This is short enough... I'm not sure I understand why it's not the exported `Root()` function?

File: ddtrace/tracer/span.go
Comment (RomainMuller) [inline ddtrace/tracer/span.go:454]: This is one more occurrence of a method that is unsafe for a `nil` receiver when everything else is... This is confusing to me (and probably a 🦶🏻🔫).

File: ddtrace/tracer/tracer.go
Comment (RomainMuller) [inline ddtrace/tracer/tracer.go:273]:
```suggestion
```

File: ddtrace/tracer/tracer.go
Comment (RomainMuller) [inline ddtrace/tracer/tracer.go:423]: Feels to me that `t.traceWriter.add` should have no side-effect if passed an empty slice, which would make this guard here unnecessary?

File: ddtrace/tracer/tracer.go
Comment (RomainMuller) [inline ddtrace/tracer/tracer.go:447]: Same as above

File: ddtrace/tracer/tracer.go
Comment (RomainMuller) [inline ddtrace/tracer/tracer.go:572]:
```suggestion
```

File: ddtrace/tracer/tracer.go
Comment (RomainMuller) [inline ddtrace/tracer/tracer.go:774]:
```suggestion
```

File: ddtrace/tracer/tracer.go
Comment (RomainMuller) [inline ddtrace/tracer/tracer.go:787]:
```suggestion
```

File: internal/appsec/_testlib/mockspan.go
Comment (RomainMuller) [inline internal/appsec/_testlib/mockspan.go:52]: Is this `MockSpan` supposed to still exist even?

File: internal/appsec/appsec.go
Comment (RomainMuller) [inline internal/appsec/appsec.go:18]: This change (and the one in `internal/appsec/config.go`) look like we might be using disjoint import formatter settings... I'd suggest we stick to `goimports` or something?

File: internal/appsec/emitter/httpsec/http.go
Comment (RomainMuller) [inline internal/appsec/emitter/httpsec/http.go:115]: This would be a synonym to `trace.TagSetter` I believe... So that's not needed?

File: internal/appsec/trace/grpctrace/grpc.go
Comment (RomainMuller) [inline internal/appsec/trace/grpctrace/grpc.go:17]: Same as above -- this issynonymous to `trace.TagSetter`

File: internal/traceprof/traceproftest/go.mod
Comment (RomainMuller) [inline internal/traceprof/traceproftest/go.mod:62]: Wouldn't a `go.work` file remove the need for this and all members of its species?

**Review by RomainMuller — COMMENTED** [review]
Love the reduction in type assertions (they always make me feel a little gross).

---

I have a bunch of API/style comments which I don't necessary feel excessively strongly about, and might actually not fit the intended scope of this PR... Generally feel free to disagree with me and/or to punt modifications to a separate/future PR (a backlog item reference would be appreciated in that case).

---

There's quite a few import ordering changes that seem entirely unrelated to the scope of this change, and which also makes me feel there might be tools fighting here in ways that automated style checks don't care about... We should probably standardize on that front (and IMHO, idiomatic tooling such as `goimports` is greatly preferable to bloated gatekeeperware such as `golangci-lint`, but I know my opinion on the subject isn't entirely objective 🫣).

To be clear - I don't mind those import changes being there (although they bloat the diff unnecessarily)... I'd just rather not see those changing back and forth with every PR depending on which IDE the author uses)

File: ddtrace/mocktracer/mockspan.go
Comment (knusbaum) [reply ddtrace/mocktracer/mockspan.go:25]: Yes, this is the new `MockSpan`. I can change it back to `MockSpan` if you prefer, but I was hoping to eliminate the stuttering `mocktracer.MockSpan`.

File: ddtrace/mocktracer/mockspan.go
Comment (knusbaum) [reply ddtrace/mocktracer/mockspan.go:88]: Probably, but not with the current baggage API. As this is a test function, I don't think the optimization is worth it.

File: ddtrace/mocktracer/mockspan.go
Comment (knusbaum) [reply ddtrace/mocktracer/mockspan.go:99]: Since it's `RLock`, we're safe from deadlocks.
However, since I've made all of the other members safe, I've gone ahead and removed the lock from this function.

File: ddtrace/mocktracer/mockspan.go
Comment (knusbaum) [reply ddtrace/mocktracer/mockspan.go:126]: `Unwrap` is probably more like what people expect. I'll change this.

File: ddtrace/mocktracer/mocktracer.go
Comment (knusbaum) [reply ddtrace/mocktracer/mocktracer.go:63]: True, but I think we should keep the explicit checks anyway so we don't accidentally stop checking that we implement the interface.

File: ddtrace/mocktracer/mocktracer.go
Comment (knusbaum) [reply ddtrace/mocktracer/mocktracer.go:77]: You're very right, this is gross.

I think this needs to be refactored. Before this was hidden behind an internal package (and was still gross, but at least internally gross). I'd like to put it back behind an internal package or unexport it or something, but at very least I'll change the signature like you suggest.

File: ddtrace/tracer/span.go
Comment (knusbaum) [reply ddtrace/tracer/span.go:225]: Yes, this should be exportable now. This is an oversight during the refactoring.

File: ddtrace/tracer/span.go
Comment (knusbaum) [reply ddtrace/tracer/span.go:454]: I've been debating. My thought was that only exported methods should have a `nil` check and internal ones should be used carefully by us, but perhaps it's a cheap enough check that we should do it everywhere to be safe.

WDYT?

File: internal/appsec/_testlib/mockspan.go
Comment (knusbaum) [reply internal/appsec/_testlib/mockspan.go:52]: Looks like it's still used, although I don't think it needs all of these methods anymore.

File: internal/appsec/appsec.go
Comment (knusbaum) [reply internal/appsec/appsec.go:18]: I defer to whatever the CI says. Agree, though, I'd be happy to use `goimports`. Let's talk about it in the guild.

File: internal/appsec/emitter/httpsec/http.go
Comment (knusbaum) [reply internal/appsec/emitter/httpsec/http.go:115]: Looks like it, yeah.

File: internal/traceprof/traceproftest/go.mod
Comment (knusbaum) [reply internal/traceprof/traceproftest/go.mod:62]: Possibly, yeah. I'll ping @darccio about that.

Comment (darccio) [reply internal/traceprof/traceproftest/go.mod:62]: @RomainMuller @knusbaum I did a quick test. A `go.work` file does help to remove the need for `replace` directives.

Let's implement it in a separate PR.

**Pull request merged**
